browser: implement native pdf rendering

deepak1556 opened the pull request:
The implementation intercepts any request to pdf resource load in main frame or sub frame (navigating to a pdf resource with `webContents.loadURL`, `window.location` apis). It will then generate a blob response of the already downloaded pdf data and redirect with it to the pdf viewer `chrome://pdf-viewer/` which is a webui page that is modified version of original extension from chromium. The pdf viewer page embeds the pepper plugin which is used to render the pdf and perform other UI actions. Currently the UI lives as a submodule at https://github.com/deepak1556/pdf-viewer. Would like feedback on the implementation, Thanks! 

- [x] blob url returns 404 when loaded from plugin
- [x] Allow download from pdf ui
- [x] Fix broken ninja rule for building `pdf_viewer_resources_map.o`
- [x] Fix failing pdf download spec
- [x] Propagate page zoom changes to the pepper plugin (Depends on https://github.com/electron/electron/pull/8537)
~~Fix loading pdf viewer for iframes (currently the parent webContents will host the viewer when sub frame tries to load a pdf resource)~~
- [x] Fix blob URL request being aborted after with chrome56 changes

Test:
- [x] Linux
- [x] Windows
- [x] Mac

Depends on:
https://github.com/electron/brightray/pull/270
https://github.com/electron/libchromiumcontent/pull/256
https://github.com/electron/libchromiumcontent/pull/263

Fixes https://github.com/electron/electron/issues/7366

![screenshot from 2017-01-18 23-13-15](https://cloud.githubusercontent.com/assets/964386/22075923/ec3a4fb8-ddd3-11e6-8fcb-185afa2c8406.png)


/cc @hokein @paulcbetts

YurySolovyov commented on atom/browser/atom_web_ui_controller_factory.cc:
```diff
@@ -0,0 +1,182 @@
+    return "text/cache-manifest";
+  }
+  return "text/html";
+}
```
This method looks like a bit ad-hoc mime types map/table implementation, it may have been solved already.

deepak1556 replied:
Fixed, thanks!

deepak1556 commented:
Thanks @hokein for testing and reviewing. Have addressed the comments.

> as discussed, would be really helpful to elaborate a few more details/steps about how electron renders PDF.

Updated the original PR description.

> I find an issue about zoom-in/out. In the page which renders the PDF, press Ctrl + +/-, the web page (toolbar) will zoom in/out, but the PDF content doesn't get zoom in/out correspondingly, which is quite different with Chrome's behavior. In chrome, only the PDF gets zoom in/out, the toolbar keeps unchanged.

Yup this is because of our current zoom apis are tied to the render process and its difficult to manage state on browser process with it.  https://github.com/electron/electron/pull/8537 addresses this problem.

deepak1556 commented:
@hokein have made the changes, thanks!

The tests are failing only on windows, seems like there is problem with loading the pdf viewer.

zcbenz commented:
How do you think of moving the `grit` step into the `pdf-viewer` repo? So we can just generate the `pdf_viewer_resources.*` files in the `pdf-viewer` repo manually, and then copy them into `electron` repo. I don't really want to include `grit` as a dependency of electron.

And in future if we were able to avoid forking `pdf-viewer`, we could just use the generated files from libchromiumcontent.

deepak1556 commented:
> How do you think of moving the grit step into the pdf-viewer repo? So we can just generate the pdf_viewer_resources.* files in the pdf-viewer repo manually, and then copy them into electron repo. I don't really want to include grit as a dependency of electron.

Sounds good.

> And in future if we were able to avoid forking pdf-viewer, we could just use the generated files from libchromiumcontent.

Yes that is very much what I would like to do in next version, the current architecture is made to work as a webui page since it was the fastest way to get a prototype pdf renderer. The way chrome does is, it embeds a browser plugin which loads the chrome extension. I am planning on going down that path in the next version of this renderer.

deepak1556 commented:
@zcbenz this is ready for review. The pdf renderer in iframe has a few issues.

* `RenderFrameHostManager::UpdateStateForNavigation` returns early for subframe navigations to use the parent render frame host which fails to trigger `RenderFrameHostManager::UpdatePendingWebUIOnCurrentFrameHost`. This will cause any navigation to pdf webui in subframe to emit `net::ERR_INVALID_URL`. This is addressed in https://codereview.chromium.org/2636193003 but will take some time for libcc to get this in.

* WebUI pages are assumed to get their own render frame host with proper binding policy set from `content/public/common/bindings_policy.h`, even though we could set this on parents' render frame host there are other places that assume the current render frame host has web ui at its root node and the ipc messages from the renderer aren't received.

Both these problems can be solved by patching one liners across a few files which is not desirable for us, the better option is to wait for OOPIF support in Electron which will fix these concerns. Thoughts on having pdf renderer support in iframes delayed ?

zcbenz commented:
Tested on Windows and it is working great!

For the `spec/fixtures/assets/pdf.pdf`, it is an example file of a PDF software, and it may not be suitable for us to include it in Electron, can you create a new PDF file instead? Exporting a PDF file from any office software or Google Docs would be fine.

And can you transfer the `pdf-viewer` repo to electron organization account?

deepak1556 commented:
@zcbenz have made the changes, thanks for testing on windows!

> can you transfer the pdf-viewer repo to electron organization account?

I don't have permission to create repo in the electron origanisation.

zcbenz commented:
> I don't have permission to create repo in the electron origanisation.

You don't need to have permission, the Transfer button in repo's Settings will send a request when you don't have permission to create repos:
<img width="754" alt="screen shot 2017-03-10 at 17 38 17" src="https://cloud.githubusercontent.com/assets/639601/23788126/8d2df58a-05b8-11e7-9a4b-826a2ba54d0f.png">

deepak1556 commented:
I have sent the transfer request, thanks!

hokein commented on atom/browser/ui/webui/pdf_viewer_handler.cc:
```diff
@@ -0,0 +1,215 @@
+void PopulateStreamInfo(base::DictionaryValue* stream_info,
+                        content::StreamInfo* stream,
+                        const std::string& original_url) {
+  std::unique_ptr<base::DictionaryValue> headers_dict(
```
nit: use `base::MakeUnique`.

hokein commented on atom/browser/ui/webui/pdf_viewer_handler.cc:
```diff
@@ -0,0 +1,215 @@
+  if (stream_) {
+    AllowJavascript();
+
+    std::unique_ptr<base::DictionaryValue> stream_info(
```
nit: base::MakeUnique

hokein commented on atom/browser/ui/webui/pdf_viewer_handler.cc:
```diff
@@ -0,0 +1,215 @@
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+
+  std::unique_ptr<base::DictionaryValue> result(new base::DictionaryValue);
```
`base::MakeUnique`

hokein commented on atom/browser/ui/webui/pdf_viewer_ui.cc:
```diff
@@ -0,0 +1,251 @@
+    content::StreamContext* stream_context =
+        content::GetStreamContextForResourceContext(resource_context);
+
+    std::unique_ptr<content::ResourceHandler> handler(
```
`base::MakeUnique`

deepak1556 commented:
@hokein have made the changes, thanks!

deepak1556 commented on atom/browser/atom_resource_dispatcher_host_delegate.cc:
```diff
@@ -95,4 +114,30 @@ AtomResourceDispatcherHostDelegate::CreateClientCertStore(
+
+void AtomResourceDispatcherHostDelegate::OnStreamCreated(
+    net::URLRequest* request,
+    std::unique_ptr<content::StreamInfo> stream) {
```
The stream created here is not passed to the pdf plugin, instead we create the stream in `PdfViewerUI`. I have removed this function.

hokein commented on atom/browser/ui/webui/pdf_viewer_handler.cc:
```diff
@@ -0,0 +1,211 @@
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+
+  if (stream_) {
```
Should we check whether `initialize_callback_id_ ` is `nullptr`? Or make a `CHECK` here.

deepak1556 replied:
Done, thanks!

zcbenz commented:
:+1:

rreimann commented:
Thx @deepak1556 for providing native pdf rendering in electron 👍

I'm wondering why the print button is currently disabled https://github.com/electron/pdf-viewer/blob/master/elements/viewer-pdf-toolbar/viewer-pdf-toolbar.html#L46 and whether it's possible to customize the provided web ui page (e.g. hiding the save button).

deepak1556 commented:
@rreimann it was intentional not implementing the print feature, was outside the scope of this PR. It is definitely possible to implement as future improvements, if need arises. As to customizing the ui, its not planned to support as part of electron but you could always provide your own `pdf_viewer_resources.pak` file as long as it implements the webui apis. If you would like to block the functionality of save button, you can use the `will-download` event of session module, all downloads in electron will go through it.

pauliusuza commented:
Toggling button visibility through config would certainly make this more useful. I have a specific case where save button needs to be hidden from view.

alexemorris commented:
Works great but I can't currently open up any customprotocol:// links in PDF's, only http://. Is there any way round this? Same problem in Chrome unless I right click link and 'open in new tab'

HemalR commented:
Trying to search high and low for documentation/examples on how to use this - does anyone have any links that might help? If native rendering is already supported, it would be ideal - otherwise will try and bundle in pdf.js and see how that goes.

Thanks in advance!

alexemorris commented:
Hi Hemal, 
Hope this helps!

```javascript
const pdfWindow = new BrowserWindow({
            width: 1024,
            height: 800,
            webPreferences: {
              plugins: true
            }
});
fileWindow.loadURL("file://PATH_TO_PDF");
```

HemalR commented:
@alexemorris Cheers buddy!

Testing it out now

HemalR commented:
Got it working, cheers again @alexemorris - slight typo in your code snippet, should read `pdfWindow` in line 8 rather than `fileWindow`.

For anybody else who may come across this, steps I am using in my app:

1. Send the pdf url via IPC from the renderer to the main process
2. Using the handler as below:

```javascript
ipcMain.on('view-pdf', (event, url) => {
	const pdfWindow = new BrowserWindow({
		width: 1024,
		height: 800,
		webPreferences: {
			plugins: true,
		},
	});
	pdfWindow.loadURL(url);
});

```

soitech commented:
Is it possible to do this with a dataUrl rather than path to a downloaded PDF? Would be nice to not have to download before displaying a generated pdf. I've tried a variety of things with no luck.

HemalR commented:
It should be - I use the exact same code for amazon object urls (where the object is a pdf) and it works as expected...
